pyquil/api/_qam.py

```diff
@@ -37,11 +42,63 @@ class QAMExecutionResult:
+    """
+
+    @property
+    def raw_readout_data(self) -> Union[RawQVMReadoutData, RawQPUReadoutData]:
+        """
+        Get the raw result data. This will be a flattened dictionary derived
+        from :class:`qcs_sdk.qvm.QVMResultData` or :class:`qcs_sdk.qpu.QPUResultData` depending on where the
+        job was run. See the respective
+
+        This property should be used when running programs that use features like
+        mid-circuit measurement and dynamic control flow on a QPU, since they can
+        produce irregular result shapes that don't necessarily fit in a
+        rectangular matrix. If the program was run on a QVM, or doesn't use those
+        features, consider using the ``register_map`` property instead.
+        """
+        return self.data.result_data.to_raw_readout_data()
 
-    execution_duration_microseconds: Optional[int] = field(default=None)
-    """Duration job held exclusive hardware access. Defaults to ``None`` when information is not available."""
+    @property
+    def register_map(self) -> Dict[str, Optional[np.ndarray]]:
+        """
+        A mapping of a register name (ie. "ro") to a ``np.ndarray`` containing the values for the
+        register.
+
+        Raises a ``RegisterMatrixConversionError`` if the inner execution data for any of the
+        registers would result in a jagged matrix. QPU result data is captured per measure,
+        meaning a value is returned for every measure to a memory reference, not just once per shot.
+        This is often the case in programs that use mid-circuit measurement or dynamic control flow,
+        where measurements to the same memory reference might occur multiple times in a shot, or be
+        skipped conditionally. In these cases, building a rectangular ``np.ndarray`` would
+        necessitate making assumptions about the data that could skew the data in undesirable ways.
+        Instead, it's recommended to manually build a matrix from the ``QPUResultData`` available
+        on the ``raw_readout_data`` property.
+        """
+        register_map = self.data.result_data.to_register_map()
+        return {key: matrix.to_ndarray() for key, matrix in register_map.items()}
+
+    @property
+    @deprecated(
+        version="4.0.0",
+        reason=(
+            "This property is ambiguous now that the `raw_readout_data` property exists"
+            "and will be removed in future versions. Use `register_map` property instead"
+        ),
+    )
+    def readout_data(self) -> Mapping[str, Optional[np.ndarray]]:
+        """Readout data returned from the QAM, keyed on the name of the readout register or post-processing node."""
+        return self.register_map
+
+    @property
+    def execution_duration_microseconds(self) -> Optional[int]:
+        """Duration job held exclusive hardware access. Defaults to ``None`` when information is not available."""
+        return self.data.duration.microseconds
```

Comment:
I'm not sure this works like you're expecting; it wraps on the second boundary: https://docs.python.org/3/library/datetime.html#datetime.timedelta

Why not just return the timedelta?

Reply:
On inspection, I need to handle the potential that `self.data.duration` could be `None`, but if it is a `timedelta`, the `microseconds` property works as expected. There is an example of it in that linked documentation.

As far why not return the `timedelta`, we could, if we wanted to make that a breaking change. Or we could deprecate it in favor of a `execution_duration` property.

Reply:
What am I missing?

```
>>> from datetime import timedelta
>>> t = timedelta(seconds=1)
>>> t.microseconds
0
```

----

> Or we could deprecate it in favor of a execution_duration property.

This sounds right, but also not necessary to do now. definitely shouldn't break the existing function to have `..._microseconds` return a `timedelta`.

Reply:
Ah, you are right, I misinterpreted my own test 🤦, sorry about that. Fixed.